### PR TITLE
Fix version comparison for versions with preRelease components

### LINKED ISSUE #53055: k8s.io/kubernetes/pkg/util/version Compare() function works incorrectly in case of prerelease builds.
**Is this a BUG REPORT or FEATURE REQUEST?**: BUG

/kind bug

**What happened**:
In some scenarios, v1.Compare(v2) would return expected result  as v2.Compare(v1)
See example code below.

```shell
$ go run kver.go
[1 8 0] alpha.3
[1 8 0] alpha.3.673 73326ef01d2d7c
1.8.0-alpha.3  AtLeast  1.8.0-alpha.3.673+73326ef01d2d7c true
1.8.0-alpha.3  LessThan 1.8.0-alpha.3.673+73326ef01d2d7c false
1.8.0-alpha.3.673+73326ef01d2d7c  AtLeast  1.8.0-alpha.3 true
1.8.0-alpha.3.673+73326ef01d2d7c  LessThan 1.8.0-alpha.3 false
1.8.0-alpha.3 Compare( 1.8.0-alpha.3.673+73326ef01d2d7c ) 0 <nil>
1.8.0-alpha.3.673+73326ef01d2d7c Compare( 1.8.0-alpha.3 ) 1 <nil>
$
```

Notice specifically:
```
1.8.0-alpha.3 Compare( 1.8.0-alpha.3.673+73326ef01d2d7c ) 0 <nil>
1.8.0-alpha.3.673+73326ef01d2d7c Compare( 1.8.0-alpha.3 ) 1 <nil>
```
So, by return value in first case it means that versions equal.

**What you expected to happen**:
```
1.8.0-alpha.3 Compare( 1.8.0-alpha.3.673+73326ef01d2d7c ) -1 <nil>
1.8.0-alpha.3.673+73326ef01d2d7c Compare( 1.8.0-alpha.3 ) 1 <nil>
```

**How to reproduce it (as minimally and precisely as possible)**:

Test program:
```go
package main

import (
	"fmt"
	"k8s.io/kubernetes/pkg/util/version"
)

func main() {
	v2 := version.MustParseSemantic("v1.8.0-alpha.3")
	v3 := version.MustParseSemantic("v1.8.0-alpha.3.673+73326ef01d2d7c")
	fmt.Println(v2.Components(), v2.PreRelease(), v2.BuildMetadata())
	fmt.Println(v3.Components(), v3.PreRelease(), v3.BuildMetadata())
	fmt.Println(v2, " AtLeast ", v3, v2.AtLeast(v3))
	fmt.Println(v2, " LessThan", v3, v2.LessThan(v3))
	fmt.Println(v3, " AtLeast ", v2, v3.AtLeast(v2))
	fmt.Println(v3, " LessThan", v2, v3.LessThan(v2))
	vvv, err := v2.Compare(v3.String())
	fmt.Println(v2, "Compare(", v3, ")", vvv, err)
	vvv, err = v3.Compare(v2.String())
	fmt.Println(v3, "Compare(", v2, ")", vvv, err)
}
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): master branch.
- Cloud provider or hardware configuration**: N/A
- OS (e.g. from /etc/os-release): N/A
- Kernel (e.g. `uname -a`): N/A
- Install tools: N/A
- Others: N/A

### PR DESCRIPTION
**Pull request by kad**
**What this PR does / why we need it**:
Fixes algorithm of comparing versions in pkg/util/version.
Also improve unit testing, so reverse comparison of versions also works

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #53055

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by kad** [comment]
cc people who contributed to original parts of that code: @danwinship @fabriziopandini

**Comment by fabriziopandini** [comment]
thanks @kad! 
CC @luxas

File: pkg/util/version/version.go [hunk 1]
```diff
@@ -181,18 +181,24 @@ func (v *Version) String() string {
...
 		case other.components[i] < v.components[i]:
 			return 1
 		case other.components[i] > v.components[i]:
 			return -1
 		}
 	}
+	// If components are common but one has more items, it is bigger
+	switch {
+	case oLen < vLen && (vLen-oLen >= 2 || v.components[oLen] != 0):
+		return 1
```
Comment (danwinship) [inline pkg/util/version/version.go]: Hm. This fixes "1.2" vs "1.2.0" but not "1.2" vs "1.2.0.0".

File: pkg/util/version/version.go [hunk 2]
```diff
@@ -234,9 +237,48 @@ func (v *Version) compareInternal(other *Version) int {
...
+	case len(oPR) < len(vPR):
+		return 1
+	case len(oPR) > len(vPR):
+		return -1
+	}
+
 	return 0
 }
 
+// Basic comparison functions
```
Comment (danwinship) [inline pkg/util/version/version.go]: This is unrelated to the bugfix (and I hadn't previously included functions like these because they weren't needed anywhere).

**Review by danwinship — CHANGES_REQUESTED** [review]

File: pkg/util/version/version.go (hunk 1 above)
Comment (kad) [reply pkg/util/version/version.go]: It's a bit weird use case, but I'll find solution today. I have idea, that I need to verify and will update PR.

File: pkg/util/version/version.go (hunk 2 above)
Comment (kad) [reply pkg/util/version/version.go]: I've got to the state where in kubeadm upgrade parts where a lot of version comparisons are happening (including selection of ranges) it would be good to have all basic comparison operators.

Comment (danwinship) [reply pkg/util/version/version.go]: OK. It should at least go in a separate commit (not one labeled "Fix version comparison for versions with preRelease components") and maybe should wait until the other PR where it's needed.

**Comment by kad** [comment]
Ok, updated PR to contain fix only and be able to detect also that `"1.2" == "1.2.0.0.0.0.0.0.0.0.0"`

**Comment by danwinship** [comment]
The `onlyZeros` is a little ugly, but I guess it works. And the extra test cases look good. LGTM.

**Review by danwinship — APPROVED** [review]

**Comment by kad** [comment]
@danwindhip if you have ideas on how to improve it to make less ugly, I'd be happy to update PR.
My other idea didn't worked out as I expected in all generic cases, thus ended up with that ugly helper func. :(

**Comment by thockin** [comment]
Can you add a test for the 1.2 == 1.2.0.0.0.0  case?

**Comment by kad** [comment]
@thockin it is there. See changed lines: 
```
 		{version: "1.9.0", unparsed: "1.9.0"},
 +		{version: "1.9.0.0.0.0.0.0", unparsed: "1.9.0.0.0.0.0.0", equalsPrev: true},
```

**Comment by luxas** [comment]
Thank you @kad!
